equinox/nn/_embedding.py

```diff
@@ -89,3 +96,313 @@ def __call__(
+              title={RoFormer: Enhanced Transformer with Rotary Position Embedding},
+              author={Jianlin Su and Yu Lu and Shengfeng Pan and Ahmed Murtadha and
+              Bo Wen and Yunfeng Liu},
+              year={2023},
+              eprint={2104.09864},
+              archivePrefix={arXiv},
+              primaryClass={cs.CL}
+            }
+        ```
+    """
+
+    embedding_size: int = field(static=True)
+
+    def __init__(
+        self,
+        embedding_size: int,
+        *,
+        key: Optional[PRNGKeyArray] = None,
+        **kwargs,
+    ):
+        """**Arguments:**
+        `RotaryPositionalEmbedding` requires:
+
+        - `embedding_size`: Size of the token embeddings. Must be non-negative.
+        - `key`: Not used; provided for compatibility with the rest of the Equinox API.
+            (Keyword only argument.)
+        """
+        if embedding_size < 0:
+            raise ValueError("embedding_size must not be negative.")
+        self.embedding_size = embedding_size
+
+    @staticmethod
+    def negate_half(x: Float[Array, "seq_len embedding_size"]):
+        d_2 = x.shape[-1] // 2
+        return jnp.concatenate([x[..., :d_2], -x[..., d_2:]], axis=-1)
+
+    @staticmethod
+    def precompute_freqs_cis(
+        embedding_size: int, end: int, theta: float = 10000.0
+    ) -> Complex[Array, "end embedding_size/2"]:
+        freqs = 1.0 / (
+            theta ** (jnp.arange(0, embedding_size, 2)[jnp.newaxis, :] / embedding_size)
+        )
+        t = jnp.arange(end)
+        freqs_outer = jnp.outer(t, freqs)
+        freqs_cis = jnp.cos(freqs_outer) + jnp.sin(freqs_outer) * 1j
+        return freqs_cis
+
+    @jax.named_scope("eqx.nn.RotaryPositionalEmbedding")
+    def __call__(
+        self,
+        x: Float[Array, "seq_len embedding_size"],
+        *,
+        key: Optional[PRNGKeyArray] = None,
+    ) -> Float[Array, "seq_len embedding_size"]:
+        """**Arguments:**
+
+        - `x`: A JAX array of shape `(seq_len, embedding_size)`.
+        - `key`: Ignored; provided for compatibility with the rest of the Equinox API.
+            (Keyword only argument.)
+
+        **Returns:**
+
+        A JAX array of shape `(seq_len, embedding_size)`, with the rotary positional
+        encoding applied to the input.
+        """
+
+        seq_len, embedding_size = x.shape
+        if embedding_size != self.embedding_size:
+            raise ValueError(
+                f"x.shape[-1] must match self.embedding_size, "
+                f"but {x.shape[-1]} != {self.embedding_size}"
+            )
+        if embedding_size % 2 != 0:
+            raise ValueError(
+                f"x.shape[-1] must be even, but {x.shape[-1]} is not even."
+            )
+
+        neg_half_x = self.negate_half(x)
+
+        with jax.ensure_compile_time_eval():
+            if (embedding_size, seq_len) in internal_rope_embedding_cache:
+                freqs_cis = internal_rope_embedding_cache[(embedding_size, seq_len)]
```

Comment:
I think we can make this even less memory-intensive: as `seq_len` increases then the cached value becomes a superset of the previous one. As such I think we should be able to keep track of precisely one array for each `embedding_size`, which represents the largest value of `seq_len` that is has seen yet. Smaller values of `seq_len` can be found by slicing this array at `[:seq_len]`.
WDYT?

Reply:
This sounds interesting. I'll build it in and check if it has any effect on performance.